webapp/javascript/components/TimelineChart/Annotations.plugin.tsx

```diff
@@ -0,0 +1,285 @@
+import React from 'react';
+import * as ReactDOM from 'react-dom';
+import Color from 'color';
+import { randomId } from '@webapp/util/randomId';
+import { CtxType } from './types';
+import extractRange from './extractRange';
+
+const AnnotationMsgIcon =
+  require('../../../images/annotationMessage.png').default;
+
+type AnnotationType = {
+  content: string;
+  timestamp: number;
+  type: 'message';
+  color: string;
+};
+
+interface IFlotOptions extends jquery.flot.plotOptions {
+  annotations?: AnnotationType[];
+  ContextMenu?: React.FC<ContextMenuProps>;
+}
+
+type AnnotationPosition = {
+  fromX: number;
+  toX: number;
+  fromY: number;
+  toY: number;
+  timestamp: number;
+  content: string;
+};
+
+export interface ContextMenuProps {
+  click: {
+    /** The X position in the window where the click originated */
+    pageX: number;
+    /** The Y position in the window where the click originated */
+    pageY: number;
+  };
+  timestamp: number;
+  containerEl: HTMLElement;
+  value?: {
+    timestamp: number;
+    content: string;
+  } | null;
+}
+
+const WRAPPER_ID = randomId('contextMenu');
+
+const getIconByAnnotationType = (type: string) => {
+  switch (type) {
+    case 'message':
+    default:
+      return AnnotationMsgIcon;
+  }
+};
+
+const shouldStartAnnotationsFunctionality = (annotations?: AnnotationType[]) =>
+  Array.isArray(annotations);
+
+const inject = ($: JQueryStatic) => {
+  const alreadyInitialized = $(`#${WRAPPER_ID}`).length > 0;
+
+  if (alreadyInitialized) {
+    return $(`#${WRAPPER_ID}`);
+  }
+
+  const body = $('body');
+  return $(`<div id="${WRAPPER_ID}" />`).appendTo(body);
+};
+
+const getCursorPositionInPx = (
+  plot: jquery.flot.plot,
+  positionInTimestamp: { x: number; y: number }
+) => {
+  const axes = plot.getAxes();
+  const extractedX = extractRange(plot, axes, 'x');
+  const extractedY = extractRange(plot, axes, 'y');
+  const plotOffset = plot.getPlotOffset() as {
+    top: number;
+    left: number;
+  };
+
+  return {
+    x: Math.floor(extractedX.axis.p2c(positionInTimestamp.x)) + plotOffset.left,
+    y: Math.floor(extractedY.axis.p2c(positionInTimestamp.y)) + plotOffset.top,
+  };
+};
+
+const findAnnotationByCursorPosition = (
+  x: number,
+  y: number,
+  list: AnnotationPosition[] = []
+) => {
+  return list?.find((an) => {
+    return x >= an.fromX && x <= an.toX && y >= an.fromY && y <= an.toY;
+  });
+};
+
+function drawRoundRect(
```

Comment:
Although I like to keep things close to where they are used, this is too much of an implementation detail. I don't want to know about canvas drawing (when reading about annotations). This and the code that draws an icon on the canvas can be moved to its own file I think.

Reply:
removed